src/visual.ts

```diff
@@ -23,1218 +23,1101 @@
 
-            const selection: UpdateSelection<StreamGraphSeries> = this.renderChart(
-                this.data.series,
-                StreamGraph.AnimationDuration
-            );
+        this.clearCatcher = appendClearCatcher(this.svg);
 
-            this.calculateAxes();
+        this.axes = this.svg
+            .append("g")
+            .classed(StreamGraph.Axes.className, true)
+            .classed(StreamGraph.axisGraphicsContext.className, true);
+        this.axisX = this.axes
+            .append("g")
+            .classed(StreamGraph.Axis.className, true)
+            .classed(StreamGraph.XAxis.className, true);
 
-            this.tooltipServiceWrapper.addTooltip(
-                selection,
-                (tooltipEvent: TooltipEventArgs<StreamGraphSeries>) => {
-                    const tooltipInfo: VisualTooltipDataItem[] = tooltipEvent.data.tooltipInfo;
+        this.axisY = this.axes
+            .append("g")
+            .classed(StreamGraph.Axis.className, true)
+            .classed(StreamGraph.YAxis.className, true);
 
-                    return tooltipInfo.length > 0
-                        ? tooltipInfo
-                        : null;
-                });
+        this.dataPointsContainer = this.svg
+            .append("g")
+            .classed(StreamGraph.DataPointsContainer, true);
 
-            const interactivityService: IInteractivityService = this.interactivityService;
+        this.behavior = new StreamGraphBehavior();
 
-            if (interactivityService) {
-                const behaviorOptions: BehaviorOptions = {
-                    selection,
-                    interactivityService,
-                    clearCatcher: this.clearCatcher
-                };
+        this.interactivityService = createInteractivityService(this.visualHost);
 
-                interactivityService.bind(
-                    this.data.series,
-                    this.behavior,
-                    behaviorOptions
-                );
+        this.legend = createLegend(
+            element,
+            false,
+            this.interactivityService,
+            true,
+        );
+    }
 
-                this.behavior.renderSelection(false);
-            }
+    public update(options: VisualUpdateOptions): void {
+        if (!options
+            || !options.dataViews
+            || !options.dataViews[0]
+            || !options.dataViews[0].categorical
+        ) {
+            this.clearData();
+            return;
         }
 
-        private setTextNodesPosition(xAxisTextNodes: Selection<any>,
-            textAnchor: string,
-            dx: string,
-            dy: string,
-            transform: string): void {
+        this.viewport = StreamGraph.getViewport(options.viewport);
+        this.dataView = options.dataViews[0];
 
-            xAxisTextNodes
-                .style("text-anchor", textAnchor)
-                .attr({ dx, dy, transform });
+        if (options.type !== 4 && options.type !== 32) {
```

Comment:
do not use 4 or 32 here:
copy interfaces:
    enum VisualUpdateType {
        Data = 2,
        Resize = 4,
        ViewMode = 8,
        Style = 16,
        ResizeEnd = 32,
        All = 62,
    }